base/asyncmap.jl

```diff
@@ -145,4 +145,27 @@ function next(itr::StreamMapIterator, state::StreamMapState)
     return (r, state)
 end
 
-iteratorsize(::Type{StreamMapIterator}) = SizeUnknown()
+iteratorsize(::Type{StreamGenerator}) = SizeUnknown()
+
+
+"""
+    asyncgenerate(f, c...) -> iterator
+
+Apply `@async f` to each element of `c`.
+
+For multiple collection arguments, apply f elementwise.
+
+Results are returned in order as they become available.
+"""
+asyncgenerate(f, c...) = StreamGenerator(f, c...)
```

Comment:
I see what you mean about `map/map!`.
I worry a little about having a type name with a bang "!" a the end. Is there any precedent for that?

As to why it is the way it is, the names `AsyncMapIterator` and `StreamMapIterator` pre-date the `Generator` name.
The former applies the mapped function with `@async`. The latter makes the results available as a stream (i.e. as they become available, rather than having to wait until the end).

I think I agree that the thing that is currently called `StreamGenerator` should probably be called `AsyncGenerator`.
I could roll these together so that the type currently called `AsyncGenerator` goes away? what do you think?